Warn for unused resource (model/seeds) config paths in project file

clrcrl opened the pull request:
Finds all model configurations specified in all projects (including installed packages) and checks whether they are applied to at least one model.
If they are not applied (i.e. "unused"), a warning is logged.
Addresses  #691

drewbanin commented:
Thanks for submitting this PR @clrcrl! We've wanted this feature for a long time -- I'm certain it will cut down on misconfigured projects immensely.

I'll add some comments here shortly, but the overall approach looks pretty good to me! Nice work :+1:

drewbanin commented on dbt/loader.py:
```diff
@@ -23,7 +24,17 @@ def load_all(cls, root_project, all_projects):
-
+        
+        model_fqns = []
+        for unique_id, node in to_return['nodes'].items():
```
You can do this as a little one-liner with a [list comprehension](http://treyhunner.com/2015/12/python-list-comprehensions-now-in-color/). It will save ~2 lines of code, but the real benefit is that the logic is more self-evident. I think the same is true for `project_model_config_pqns` below

clrcrl commented on dbt/loader.py:
```diff
@@ -23,7 +24,17 @@ def load_all(cls, root_project, all_projects):
 
             to_return[subgraph] = subgraph_nodes
-
+        
```
Assume you meant to write "`compilation.py` file"?

drewbanin replied:
yeah, that's right, good catch :)

clrcrl commented on dbt/loader.py:
```diff
@@ -23,7 +24,17 @@ def load_all(cls, root_project, all_projects):
+        
+        project_model_config_pqns = []
+        for project_name, project in all_projects.items():
+            project_model_config_pqns.extend(dbt.config.get_project_model_config_pqns(project))
```
I've updated this bit of code to only read the `root_project` instead of `all_projects`. 

I think the helper function is still required because it's un-nesting the config into the list of lists.

clrcrl commented on dbt/config.py:
```diff
@@ -43,3 +44,57 @@ def send_anonymous_usage_stats(config):
+    unused_pqns = get_unused_pqns(pqns, fqns)
+    if len(unused_pqns) == 0:
+        return
+    logger.info(dbt.ui.printer.yellow("WARNING: Model configurations exist in your dbt_project.yml file which do not apply to any models. There are {} unused model configurations:".format(len(unused_pqns))))
```
Probably a separate issue/PR, as I'd also want to have `logger.debug` print in red text, and then go back and clean up where the color is being directly called ([example](https://github.com/fishtown-analytics/dbt/blob/development/dbt/node_runners.py#L122)), and possibly remove green/yellow/red functions from [printer.py](https://github.com/fishtown-analytics/dbt/blob/development/dbt/ui/printer.py)

clrcrl commented:
@drewbanin - addressed most of your comments.
Agreed on the recursive function front. Even if we can't figure out a way to implement it without mutating the `pqns` argument, it still feels like it can be tidied up! (Especially since codeclimate is complaining about it)

clrcrl commented:
To do:
Add unit tests [here](https://github.com/fishtown-analytics/dbt/blob/development/test/unit/test_config.py), e..g `test__warning_for_unused_configs`:
* a positive test, i.e. one test that returns 0 unused configs
* a negative test, i.e. one test that returns > 0 unused configs

clrcrl commented:
Another few To-Do:
* Correct grammar for warning for a single error: "A model configuration exists ... There is 1 unused model configuration" 
* When there are > 4 models, warn about first three and X others
e.g.`
```
WARNING: Model configurations exist in your dbt_project.yml file which do not apply to any models. There are 6 unused model configurations:
 - experiments.base
 - experiments.tables
 - experiments.views
 and 3 others
```

This error occurred because I had:
```
models:
  my_package_name:
    ...
  experiments: #not indented enough
    base:
      ...
    tables:
      ...
    views:
      ...
```
instead of
```
models:
  my_package_name:
    ...
    experiments:
      base:
        ...
      tables:
        ...
      views:
        ...
      ...
```
Actually a thought here - if the config is wrong because of a high level nesting issue, do we just want to warn about that level, rather than every single pqn. Especially since it's the one mistake that has affected multiple configs.
```
WARNING: Model configuration(s) exist in your dbt_project.yml file which do not apply to any models. There is 1 unused model configuration:
 - experiments
```

clrcrl commented:
^ ended up not implementing the above to-dos around making the warning message friendlier. Can do if you think it is valuable.

That recursive function still probably needs some love.

Over to you @drewbanin

clrcrl commented:
Now addresses #784.
Keen to get feedback on this!

clrcrl commented:
@drewbanin – couldn't get the integration tests running locally (docker issues), but used this to check my tests:
```
(dbt-dev) $ python -m unittest -q test/unit/test_config.py
----------------------------------------------------------------------
Ran 6 tests in 0.022s

OK
```
Expecting it to fail codeclimate again though

clrcrl commented:
Nope - don't look at this yet. Just realised there's an error in it

clrcrl commented:
OK should be working now.
I think I need to write some better tests though. I feel like it might be a good opportunity to pair for this part since this is part of writing code that I want to get a whole lot better at.

drewbanin commented:
Fixed in https://github.com/fishtown-analytics/dbt/pull/1027 -- closing!
